provider/cmd/crd2pulumi/generate.go

```diff
@@ -0,0 +1,209 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/pulumi/pulumi-kubernetes/provider/cmd/crd2pulumi/nodejs"
+	unstruct "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+const (
+	DotNet string = "dotnet"
+	Go     string = "go"
+	NodeJS string = "nodejs"
+	Python string = "python"
+)
+
+const (
+	v1beta1 string = "apiextensions.k8s.io/v1beta1"
+	v1      string = "apiextensions.k8s.io/v1"
+)
+
+type CustomResourceGenerator struct {
+	// CustomResourceDefinition represents unmarshalled CRD YAML
+	CustomResourceDefinition unstruct.Unstructured
+	// OutputPath represents the path of the file to generate code in
+	OutputPath string
+	// Language represents the target language to generate code
+	Language string
+	// ApiVersion represents the `apiVersion` field in the CRD YAML
+	APIVersion string
+	// Kind represents the `spec.names.kind` field in the CRD YAML
+	Kind string
+	// Plural represents the `spec.names.plural` field in the CRD YAML
+	Plural string
+	// Group represents the `spec.group` field in the CRD YAML
+	Group string
+	// Versions represents a mapping from each version name in the
+	// `spec.versions` list to its corresponding `openAPIV3Schema` field in the
+	// CRD YAML
+	Versions map[string]map[string]interface{}
+}
+
+func NewCustomResourceGenerator(language, yamlPath, outputPath string) (CustomResourceGenerator, error) {
+	if language != DotNet && language != Go && language != NodeJS && language != Python {
+		return CustomResourceGenerator{}, errors.New("invalid language: " + language)
+	}
+
+	yamlFile, err := ioutil.ReadFile(yamlPath)
+	if err != nil {
+		return CustomResourceGenerator{}, fmt.Errorf("read file %s: %v", yamlPath, err)
```

Comment:
Consider importing [`"github.com/pkg/errors"`](https://github.com/pkg/errors) and wrap the error like:

```suggestion
		return CustomResourceGenerator{}, errors.Wrapf(err, "reading file: %s", yamlPath)
```

Applies throughout